infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntime.java

```diff
@@ -139,6 +130,48 @@ protected void internalStart(Map<String, String> startOptions) throws Infrastruc
     }
   }
 
+  /**
+   * Bootstraps machines in parallel.
+   *
+   * @throws InfrastructureException when any error occurs while bootstrapping machines
+   * @throws InterruptedException when machines bootstrapping was interrupted
+   */
+  private void bootstrapMachines() throws InfrastructureException, InterruptedException {
+    List<CompletableFuture> checkMachines = new ArrayList<>(machines.size());
+    final ExecutorService executorService = Executors.newCachedThreadPool();
+    for (OpenShiftMachine machine : machines.values()) {
+      checkMachines.add(
+          CompletableFuture.runAsync(
+                  () -> {
+                    try {
+                      machine.waitRunning(machineStartTimeoutMin);
+                      bootstrapMachine(machine);
+                      checkMachineServers(machine);
+                    } catch (InfrastructureException ex) {
+                      executorService.shutdownNow();
+                      sendFailedEvent(machine.getName(), ex.getMessage());
+                      throw new CompletionException(ex);
+                    } catch (InterruptedException ei) {
+                      executorService.shutdownNow();
+                      throw new CompletionException(ei);
+                    }
+                  },
+                  executorService)
+              .thenAccept(a -> sendRunningEvent(machine.getName())));
```

Comment:
Are you sure that you need setting `thenAccept` here? Looks like you can put sending running event just after servers checking line. I mean 
```
try {
  machine.waitRunning(machineStartTimeoutMin);
  bootstrapMachine(machine);
  checkMachineServers(machine);
  sendRunningEvent(machine.getName());
} catch ...
```

Reply:
Makes sense